<h2 align='center' style='color:blue'>Finding best model and hyper parameters for sklearn digits dataset classification

**for digits dataset in sklearn.dataset,please try the following classifiers and find out the one that gives best performance.Also find the optimal parameters for that classifier**

In [1]:
from sklearn.svm import SVC
from sklearn.ensemble import RandomForestClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.naive_bayes import GaussianNB,MultinomialNB
from sklearn.linear_model import LogisticRegression



In [2]:
from sklearn.datasets import load_digits

digits=load_digits()

In [3]:
model_params={
    'SVM':{
        'model':SVC(),
        'params':{
            'C':[1,10,10],
            'kernel':['linear', 'poly', 'rbf', 'sigmoid'],
            'gamma':["auto"]
        }
    },
    'random_forest':{
        'model':RandomForestClassifier(),
        'params':{
            'n_estimators':[1,5,10]
        }
    },
    'decision_tree':{
        'model':DecisionTreeClassifier(),
        'params':{
            "criterion":['gini', 'entropy', 'log_loss']
        }
    },
    'gaussian_NB':{
        'model':GaussianNB(),
        'params':{
            
        }
    },
    'Multinomial_NB':{
        'model':MultinomialNB(),
        'params':{}
    },
    'logistic_regression':{
        'model':LogisticRegression(),
        'params':{
            'C':[1,5,10]
        }
    }

}

In [4]:
from sklearn.model_selection import GridSearchCV

In [5]:
scores=[]

for model_name,mp in model_params.items():
    clf=GridSearchCV(estimator=mp['model'],param_grid=mp['params'],cv=5,return_train_score=False)
    clf.fit(digits.data,digits.target)
    scores.append({
        "model":model_name,
        "best_score":clf.best_score_,
        "best_parameter":clf.best_params_
    })

/Users/kristalshrestha/Documents/Code/LLM_Scratch/venv/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:473: ConvergenceWarning: lbfgs failed to converge after 100 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=100).
You might also want to scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(
/Users/kristalshrestha/Documents/Code/LLM_Scratch/venv/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:473: ConvergenceWarning: lbfgs failed to converge after 100 iteration(s) (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT

Increase the number of iterations to improve the convergence (max_iter=100).
You might also want to scale the data as shown

In [9]:
import pandas as pd

df=pd.DataFrame(scores)
df

,model,best_score,best_parameter
0,SVM,0.968842,"{'C': 1, 'gamma': 'auto', 'kernel': 'poly'}"
1,random_forest,0.897640,{'n_estimators': 10}
2,decision_tree,0.812501,{'criterion': 'entropy'}
3,gaussian_NB,0.806928,{}
4,Multinomial_NB,0.870350,{}
5,logistic_regression,0.913765,{'C': 1}


**For me the winner is svm (C=1, kernel=poly,gamma=auto) with 96.88% score. It could be different for you as I have limited my parameters to be certain values only**